[1.21.1] Replace "No key duration in MapLike..." with clearer error message

### Checked for existing issues

- [x] I have checked for existing issues, and have found none.

### Tested latest version

- [x] I have checked that this occurs on the latest version.

### GregTech CEu Version

reab/1.21-with-love

### Minecraft Version

1.21.1 NeoForge

### Recipe Viewer Installed

None

### Environment

Singleplayer

### Cross-Mod Interaction

Yes

### Other Installed Mods

dev env

### Expected Behavior

When making a recipe that errors for any reason, it should clearly say that the recipe is invalid

### Actual Behavior


```
[14:51:20] [ERROR] ! server_scripts:example.js#13: Error parsing recipe gtceu:assembler/test1 (details below this line): com.google.gson.JsonParseException: No key duration in MapLike[{"type":"gtceu:assembler","_kubejs_changed_marker":{"source":"server_scripts:example.js","line":13}}]
[14:51:20] [ERROR] ! com.google.gson.JsonParseException: No key duration in MapLike[{"type":"gtceu:assembler","_kubejs_changed_marker":{"source":"server_scripts:example.js","line":13}}]
```

### Steps to Reproduce

```
ServerEvents.recipes(event => {
    event.recipes.gtceu.assembler("test1")
        .itemInputs('2x gtceu:iron_geaaaar', 'gtceu:iron_plate')
        .itemOutputs('8x minecraft:dirt')
        .circuit(11)
        .duration(50)
        .EUt(8);
})
```

### Additional Information

It does also say above "[14:51:20] [ERROR] ! server_scripts:example.js#13: Error in 'ServerEvents.recipes': Failed to read ingredient from 2x gtceu:iron_geaaaar: Item with ID gtceu:iron_geaaaar does not exist!" so it's not the highest prio, just thought it'd be something neat to make logging more verbose

This is the full error btw, for debugging:
```

[14:51:20] [ERROR] ! server_scripts:example.js#13: Error parsing recipe gtceu:assembler/test1 (details below this line): com.google.gson.JsonParseException: No key duration in MapLike[{"type":"gtceu:assembler","_kubejs_changed_marker":{"source":"server_scripts:example.js","line":13}}]
[14:51:20] [ERROR] ! com.google.gson.JsonParseException: No key duration in MapLike[{"type":"gtceu:assembler","_kubejs_changed_marker":{"source":"server_scripts:example.js","line":13}}]
[14:51:20] [ERROR] !   at com.mojang.serialization.DataResult$Error.getOrThrow(DataResult.java:287)
[14:51:20] [ERROR] !   at net.minecraft.world.item.crafting.RecipeManager.apply(RecipeManager.java:60)
[14:51:20] [ERROR] !   at net.minecraft.world.item.crafting.RecipeManager.apply(RecipeManager.java:36)
[14:51:20] [ERROR] !   at net.minecraft.server.packs.resources.SimplePreparableReloadListener.lambda$reload$1(SimplePreparableReloadListener.java:19)
[14:51:20] [ERROR] !   at java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:718)
[14:51:20] [ERROR] !   at java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:482)
[14:51:20] [ERROR] !   at net.minecraft.server.packs.resources.SimpleReloadInstance.lambda$new$3(SimpleReloadInstance.java:86)
[14:51:20] [ERROR] !   at net.minecraft.server.TickTask.run(TickTask.java:18)
[14:51:20] [ERROR] !   at net.minecraft.util.thread.BlockableEventLoop.doRunTask(BlockableEventLoop.java:151)
[14:51:20] [ERROR] !   at net.minecraft.util.thread.ReentrantBlockableEventLoop.doRunTask(ReentrantBlockableEventLoop.java:24)
[14:51:20] [ERROR] !   at net.minecraft.server.MinecraftServer.doRunTask(MinecraftServer.java:881)
[14:51:20] [ERROR] !   at net.minecraft.server.MinecraftServer.doRunTask(MinecraftServer.java:170)
[14:51:20] [ERROR] !   at net.minecraft.util.thread.BlockableEventLoop.pollTask(BlockableEventLoop.java:122)
[14:51:20] [ERROR] !   at net.minecraft.server.MinecraftServer.pollTaskInternal(MinecraftServer.java:864)
[14:51:20] [ERROR] !   at net.minecraft.server.MinecraftServer.pollTask(MinecraftServer.java:858)
[14:51:20] [ERROR] !   at net.minecraft.util.thread.BlockableEventLoop.managedBlock(BlockableEventLoop.java:135)
[14:51:20] [ERROR] !   at net.minecraft.server.MinecraftServer.managedBlock(MinecraftServer.java:830)
[14:51:20] [ERROR] !   at net.minecraft.server.MinecraftServer.reloadResources(MinecraftServer.java:1548)
[14:51:20] [ERROR] !   at net.minecraft.server.commands.ReloadCommand.reloadPacks(ReloadCommand.java:20)
[14:51:20] [ERROR] !   at net.minecraft.server.commands.ReloadCommand.lambda$register$3(ReloadCommand.java:54)
[14:51:20] [ERROR] !   at com.mojang.brigadier.context.ContextChain.runExecutable(ContextChain.java:73)
[14:51:20] [ERROR] !   at net.minecraft.commands.execution.tasks.ExecuteCommand.execute(ExecuteCommand.java:29)
[14:51:20] [ERROR] !   at net.minecraft.commands.execution.tasks.ExecuteCommand.execute(ExecuteCommand.java:13)
[14:51:20] [ERROR] !   at net.minecraft.commands.execution.UnboundEntryAction.lambda$bind$0(UnboundEntryAction.java:8)
[14:51:20] [ERROR] !   at net.minecraft.commands.execution.CommandQueueEntry.execute(CommandQueueEntry.java:5)
[14:51:20] [ERROR] !   at net.minecraft.commands.execution.ExecutionContext.runCommandQueue(ExecutionContext.java:103)
[14:51:20] [ERROR] !   at net.minecraft.commands.Commands.executeCommandInContext(Commands.java:363)
[14:51:20] [ERROR] !   at net.minecraft.commands.Commands.performCommand(Commands.java:284)
[14:51:20] [ERROR] !   at net.minecraft.server.network.ServerGamePacketListenerImpl.performUnsignedChatCommand(ServerGamePacketListenerImpl.java:1312)
[14:51:20] [ERROR] !   at net.minecraft.server.network.ServerGamePacketListenerImpl.lambda$handleChatCommand$7(ServerGamePacketListenerImpl.java:1299)
[14:51:20] [ERROR] !   at net.minecraft.server.TickTask.run(TickTask.java:18)
[14:51:20] [ERROR] !   at net.minecraft.util.thread.BlockableEventLoop.doRunTask(BlockableEventLoop.java:151)
[14:51:20] [ERROR] !   at net.minecraft.util.thread.ReentrantBlockableEventLoop.doRunTask(ReentrantBlockableEventLoop.java:24)
[14:51:20] [ERROR] !   at net.minecraft.server.MinecraftServer.doRunTask(MinecraftServer.java:881)
[14:51:20] [ERROR] !   at net.minecraft.server.MinecraftServer.doRunTask(MinecraftServer.java:170)
[14:51:20] [ERROR] !   at net.minecraft.util.thread.BlockableEventLoop.pollTask(BlockableEventLoop.java:122)
[14:51:20] [ERROR] !   at net.minecraft.server.MinecraftServer.pollTaskInternal(MinecraftServer.java:864)
[14:51:20] [ERROR] !   at net.minecraft.server.MinecraftServer.pollTask(MinecraftServer.java:858)
[14:51:20] [ERROR] !   at net.minecraft.util.thread.BlockableEventLoop.managedBlock(BlockableEventLoop.java:135)
[14:51:20] [ERROR] !   at net.minecraft.server.MinecraftServer.managedBlock(MinecraftServer.java:830)
[14:51:20] [ERROR] !   at net.minecraft.server.MinecraftServer.mixinextras$bridge$managedBlock$278(MinecraftServer.java)
[14:51:20] [ERROR] !   at net.minecraft.server.MinecraftServer.wrapOperation$zgo000$modernfix$managedBlock(MinecraftServer.java:2181)
[14:51:20] [ERROR] !   at net.minecraft.server.MinecraftServer.waitUntilNextTick(MinecraftServer.java:835)
[14:51:20] [ERROR] !   at net.minecraft.server.MinecraftServer.runServer(MinecraftServer.java:718)
[14:51:20] [ERROR] !   at net.minecraft.server.MinecraftServer.lambda$spin$2(MinecraftServer.java:267)
[14:51:20] [ERROR] !   at java.lang.Thread.run(Thread.java:1583)
[14:51:20] [ERROR] ! server_scripts:example.js#13: Recipe JSON for gtceu:assembler/test1: {"type":"gtceu:assembler"}

```

--- a/src/main/java/com/gregtechceu/gtceu/api/codec/GTCodecUtils.java
+++ b/src/main/java/com/gregtechceu/gtceu/api/codec/GTCodecUtils.java
@@ -1,8 +1,12 @@
 package com.gregtechceu.gtceu.api.codec;
 
+import com.gregtechceu.gtceu.GTCEu;
+
+import com.google.gson.JsonParseException;
 import com.mojang.datafixers.util.Either;
 import com.mojang.serialization.Codec;
 import com.mojang.serialization.DataResult;
+import com.mojang.serialization.MapCodec;
 
 import java.util.function.Function;
 
@@ -29,5 +33,27 @@ public class GTCodecUtils {
 
     public static <T> T unboxEither(Either<T, T> either) {
         return either.map(Function.identity(), Function.identity());
+    }
+
+    public static <T> MapCodec<T> quietExceptionCodec(Codec<T> codec, String field, boolean isKubeLoaded) {
+        return codec.optionalFieldOf(field, null).flatXmap(
+                val -> {
+                    if (val != null) return DataResult.success(val);
+
+                    String msg = "Recipe " + field + " field is invalid!";
+                    if (isKubeLoaded) {
+                        throw quietException(msg);
+                    } else {
+                        GTCEu.LOGGER.error(msg);
+                    }
+                    return DataResult.error(() -> msg);
+                },
+                DataResult::success);
+    }
+
+    public static JsonParseException quietException(String msg) {
+        var ex = new JsonParseException(msg);
+        ex.setStackTrace(new StackTraceElement[0]);
+        return ex;
     }
 }

--- a/src/main/java/com/gregtechceu/gtceu/api/recipe/GTRecipeSerializer.java
+++ b/src/main/java/com/gregtechceu/gtceu/api/recipe/GTRecipeSerializer.java
@@ -294,9 +294,7 @@ public class GTRecipeSerializer implements RecipeSerializer<GTRecipe> {
 
                     String msg = "Recipe " + field + " field is invalid!";
                     if (isKubeLoaded) {
-                        var ex = new com.google.gson.JsonParseException(msg);
-                        ex.setStackTrace(new StackTraceElement[0]);
-                        throw ex;
+                        throw quietException(msg);
                     } else {
                         GTCEu.LOGGER.error(msg);
                     }

--- a/src/main/java/com/gregtechceu/gtceu/api/recipe/GTRecipeSerializer.java
+++ b/src/main/java/com/gregtechceu/gtceu/api/recipe/GTRecipeSerializer.java
@@ -24,6 +24,7 @@ import net.minecraft.util.Tuple;
 import net.minecraft.world.item.crafting.RecipeSerializer;
 import net.minecraft.world.item.crafting.RecipeType;
 
+import com.google.gson.JsonParseException;
 import com.mojang.serialization.Codec;
 import com.mojang.serialization.DataResult;
 import com.mojang.serialization.MapCodec;
@@ -262,7 +263,7 @@ public class GTRecipeSerializer implements RecipeSerializer<GTRecipe> {
                     CHANCE_LOGIC_MAP_CODEC.optionalFieldOf("tickOutputChanceLogics", Map.of()).forGetter(val -> val.tickOutputChanceLogics),
                     RecipeCondition.CODEC.listOf().optionalFieldOf("recipeConditions", List.of()).forGetter(val -> val.conditions),
                     CompoundTag.CODEC.optionalFieldOf("data", new CompoundTag()).forGetter(val -> val.data),
-                    ExtraCodecs.NON_NEGATIVE_INT.fieldOf("duration").forGetter(val -> val.duration),
+                    quietExceptionCodec(ExtraCodecs.NON_NEGATIVE_INT,"duration",isKubeLoaded).forGetter(val -> val.duration),
                     GTRegistries.RECIPE_CATEGORIES.byNameCodec().optionalFieldOf("category", GTRecipeCategory.DEFAULT).forGetter(val -> val.recipeCategory)
             ).apply(instance, GTRecipe::new));
         } else {
@@ -279,11 +280,35 @@ public class GTRecipeSerializer implements RecipeSerializer<GTRecipe> {
                     RecipeCondition.CODEC.listOf().optionalFieldOf("recipeConditions", List.of()).forGetter(val -> val.conditions),
                     IngredientActionHolder.CODEC.listOf().optionalFieldOf("kubejs:actions", List.of()).forGetter(val -> (List<IngredientActionHolder>) val.ingredientActions),
                     CompoundTag.CODEC.optionalFieldOf("data", new CompoundTag()).forGetter(val -> val.data),
-                    ExtraCodecs.NON_NEGATIVE_INT.fieldOf("duration").forGetter(val -> val.duration),
+                    quietExceptionCodec(ExtraCodecs.NON_NEGATIVE_INT,"duration",isKubeLoaded).forGetter(val -> val.duration),
                     GTRegistries.RECIPE_CATEGORIES.byNameCodec().optionalFieldOf("category", GTRecipeCategory.DEFAULT).forGetter(val -> val.recipeCategory)
             ).apply(instance, GTRecipe::new));
         }
         // spotless:on
+    }
+
+    private static <T> MapCodec<T> quietExceptionCodec(Codec<T> codec, String field, boolean isKubeLoaded) {
+        return codec.optionalFieldOf(field, null).flatXmap(
+                val -> {
+                    if (val != null) return DataResult.success(val);
+
+                    String msg = "Recipe " + field + " field is invalid!";
+                    if (isKubeLoaded) {
+                        var ex = new com.google.gson.JsonParseException(msg);
+                        ex.setStackTrace(new StackTraceElement[0]);
+                        throw ex;
+                    } else {
+                        GTCEu.LOGGER.error(msg);
+                    }
+                    return DataResult.error(() -> msg);
+                },
+                DataResult::success);
+    }
+
+    private static JsonParseException quietException(String msg) {
+        var ex = new JsonParseException(msg);
+        ex.setStackTrace(new StackTraceElement[0]);
+        return ex;
     }
 
     public static class KJSCallWrapper {

--- a/src/main/java/com/gregtechceu/gtceu/api/recipe/GTRecipeSerializer.java
+++ b/src/main/java/com/gregtechceu/gtceu/api/recipe/GTRecipeSerializer.java
@@ -24,7 +24,6 @@ import net.minecraft.util.Tuple;
 import net.minecraft.world.item.crafting.RecipeSerializer;
 import net.minecraft.world.item.crafting.RecipeType;
 
-import com.google.gson.JsonParseException;
 import com.mojang.serialization.Codec;
 import com.mojang.serialization.DataResult;
 import com.mojang.serialization.MapCodec;
@@ -36,6 +35,8 @@ import org.jetbrains.annotations.NotNull;
 
 import java.util.*;
 import java.util.function.Function;
+
+import static com.gregtechceu.gtceu.api.codec.GTCodecUtils.quietExceptionCodec;
 
 @SuppressWarnings("DataFlowIssue")
 public class GTRecipeSerializer implements RecipeSerializer<GTRecipe> {
@@ -285,28 +286,6 @@ public class GTRecipeSerializer implements RecipeSerializer<GTRecipe> {
             ).apply(instance, GTRecipe::new));
         }
         // spotless:on
-    }
-
-    private static <T> MapCodec<T> quietExceptionCodec(Codec<T> codec, String field, boolean isKubeLoaded) {
-        return codec.optionalFieldOf(field, null).flatXmap(
-                val -> {
-                    if (val != null) return DataResult.success(val);
-
-                    String msg = "Recipe " + field + " field is invalid!";
-                    if (isKubeLoaded) {
-                        throw quietException(msg);
-                    } else {
-                        GTCEu.LOGGER.error(msg);
-                    }
-                    return DataResult.error(() -> msg);
-                },
-                DataResult::success);
-    }
-
-    private static JsonParseException quietException(String msg) {
-        var ex = new JsonParseException(msg);
-        ex.setStackTrace(new StackTraceElement[0]);
-        return ex;
     }
 
     public static class KJSCallWrapper {

